src/jdk.jdwp.agent/unix/native/libjdwp/exec_md.c

```diff
@@ -44,6 +47,99 @@ static char *skipNonWhitespace(char *p) {
     return p;
 }
 
+int
+isAsciiDigit(char c)
+{
+  return c >= '0' && c <= '9';
+}
+
+#if defined(_AIX)
+  /* AIX does not understand '/proc/self' - it requires the real process ID */
+  #define FD_DIR aix_fd_dir
+  #define DIR DIR64
+  #define dirent dirent64
+  #define opendir opendir64
+  #define readdir readdir64
+  #define closedir closedir64
+#elif defined(_ALLBSD_SOURCE)
+  #define FD_DIR "/dev/fd"
+#else
+  #define FD_DIR "/proc/self/fd"
+#endif
+
+// closes every file descriptor that is listed as a directory
+// entry in "/proc/self/fd" (or its equivalent). standard
+// input/output/error file descriptors will not be closed
+// by this function. this function returns 0 on failure
+// and 1 on success
+int
+closeDescriptors(void)
+{
+    DIR *dp;
+    struct dirent *dirp;
+    /* leave out standard input/output/error descriptors */
+    int from_fd = STDERR_FILENO + 1;
+
+    /* We're trying to close all file descriptors, but opendir() might
+     * itself be implemented using a file descriptor, and we certainly
+     * don't want to close that while it's in use.  We assume that if
+     * opendir() is implemented using a file descriptor, then it uses
+     * the lowest numbered file descriptor, just like open().  So we
+     * close a couple explicitly, so that opendir() can then use
+     * these lowest numbered closed file descriptors afresh.   */
+
+    close(from_fd);          /* for possible use by opendir() */
+    close(from_fd + 1);      /* another one for good luck */
+    from_fd += 2; /* leave out the 2 we just closed, which the opendir() may use */
+
+#if defined(_AIX)
+    /* set FD_DIR for AIX which does not understand '/proc/self' - it
+     * requires the real process ID */
+    char aix_fd_dir[32];     /* the pid has at most 19 digits */
+    snprintf(aix_fd_dir, 32, "/proc/%d/fd", getpid());
+#endif
+
+    if ((dp = opendir(FD_DIR)) == NULL)
+        return 0; // failure
+
+    while ((dirp = readdir(dp)) != NULL) {
+        int fd;
+        if (isAsciiDigit(dirp->d_name[0]) &&
+            (fd = strtol(dirp->d_name, NULL, 10)) >= from_fd)
+            close(fd);
+    }
+
+    closedir(dp);
+
+    return 1; // success
+}
+
+// does necessary housekeeping of a forked child process
+// (like closing copied file descriptors) before
+// execing the child process. this function never returns
+void
+forkedChildProcess(const char *file, char *const argv[])
+{
+    /* close all file descriptors that have been copied over
+     * from the parent process due to fork() */
+    if (closeDescriptors() == 0) { /* failed,  close the old way */
+        /* find max allowed file descriptors for a process
+         * and assume all were opened for the parent process and
+         * copied over to this child process. we close them all */
+        const rlim_t max_fd = sysconf(_SC_OPEN_MAX);
```

Comment:
Why not use `int`, like the original code, instead of  `rlim_t` - as per man page `close()` expects `int`, not  `rlim_t`, ex:

```
        const int max_fd = (int)sysconf(_SC_OPEN_MAX);
        JDI_ASSERT(max_fd != -1);
        int fd;
         /* leave out standard input/output/error file descriptors */
         for (fd = STDERR_FILENO + 1; fd < max_fd; fd++) {
             (void)close(fd);
         }
```

Reply:
Hello Gerard, my understanding is that the limit value configured may exceed the int range. I wanted to avoid the overflow by casting it to int in such cases. I had noticed close() takes an int, but I couldn't think of any other way of avoiding the overflow at this place.

In the JVM parent process we do however limit it to INT_MAX. So maybe I should assume that it will be an int cast it to int here, like you suggest, and add a code comment explaining this? Does that sound OK?

Reply:
That is fine, but in that case we should also do:

`JDI_ASSERT(max_fd <= INT_MAX);`

in case `sysconf(_SC_OPEN_MAX)` returns value greater than `INT_MAX`, since `close()` only accepts `int`.

Reply:
Done. I've updated the PR to add this assert. Additionally I've also slightly changed the implementation in the newly introduced closeDescriptors() method to ensure that we don't pass values to close() that exceed INT_MAX.